### PR TITLE
room state update

### PR DESCRIPTION
**Pull request by zt4ff**
The Room class used to have two data types for the rooms state with similar methods working on the state.

I merged both states because the method operating on them are almost similar

### CODE REVIEW

File: server/__tests__/websocket.test.ts [hunk 1]
```diff
@@ -41,15 +45,19 @@ describe("User", () => {
     // creates four users
     for (let i = 0; i <= 3; i++) {
       users[i] = await new RandomUser().join(room);
+      console.log(users[i]);
```
Comment (Skrpk) [inline server/__tests__/websocket.test.ts]: Do we need this log in "prod"?

File: server/game_websocket/index.ts [hunk 2]
```diff
@@ -12,7 +12,8 @@ export const socketListener = async (server: http.Server, room: Room) => {
 
     // the rooms of the user is created asynchronous
     const user = await new RandomUser().join(room);
-    // const user = { name: "boy" };
+    console.log(user);
```
Comment (Skrpk) [inline server/game_websocket/index.ts]: Same

File: server/game_websocket/room.ts [hunk 3]
```diff
@@ -1,71 +1,52 @@
...
   // contains the state of the rooms (number of users, id)
-  private roomsDetails: Array<Rooms>;
+  private roomState: Array<Rooms>;
   // keeps tab of users added in new rooms
-  private roomsState: {
-    [key: string]: Array<{ username: string; color: string }>;
-  };
+  // private roomsState: {
+  //   [key: string]: Array<{ username: string; color: string }>;
+  // };
```
Comment (Skrpk) [inline server/game_websocket/room.ts]: Do we need it?

File: server/__tests__/websocket.test.ts [hunk 4]
```diff
@@ -41,15 +45,19 @@ describe("User", () => {
...
   it("confirms the number of users created", () => {
     expect(room.getOnlineUsers()).toBe(4);
   });
 
-  it("", async () => {
-    console.log(room.getUsersInRoom(users[0]));
-    // expect(room.getUsersInRoom(users[0]).length).toBe(3);
+  it("add all users to list of players in a room", async () => {
+    for (let i = 0; i < users.length; i++) {
+      room.addUserToRoom(users[i]);
```
Comment (Skrpk) [inline server/__tests__/websocket.test.ts]: I launched this test locally and it failed with an errors:
`Property 'addUserToRoom' does not exist on type 'Room'`
`Property 'removeUserFromRoom' does not exist on type 'Room'.`
And it's right, these properties doesn't exist on Room type. Is it ok for now?

File: server/game_websocket/room.ts [hunk 5]
```diff
@@ -76,13 +57,24 @@ export class Room {
     });
   }
 
+  public getUsersInRoom(user: User) {
+    // return this.roomsState[user.roomID];
+    const room = this.roomState.filter((room) => {
+      if (room.id === user.roomID) return true;
+    });
+    return room[0].users;
+    console.log(room[0].users);
```
Comment (Skrpk) [inline server/game_websocket/room.ts]: Unreachable code + do we need this log here?

File: server/game_websocket/room.ts [hunk 6]
```diff
@@ -1,71 +1,52 @@
 // this is where the rooms are managed and created
 import { v4 as uuidv4 } from "uuid";
-type Rooms = { users: number; id: string; board: Board };
+type Rooms = {
```
Comment (Skrpk) [inline server/game_websocket/room.ts]: Current naming with rooms looks a bit confusing. Lets name it not Rooms, but`roomState` and then rename `roomState` property of `Room` class to `roomsState`.

File: server/game_websocket/room.ts [hunk 7]
```diff
@@ -1,71 +1,52 @@
...
+  users: Array<{ username: string; color: string }>;
+  id: string;
+  board: Board;
+};
 import { Board } from "./board";
+import { RandomUser } from "./user";
```
Comment (Skrpk) [inline server/game_websocket/room.ts]: Unused variable, you can remove this import

File: server/game_websocket/room.ts [hunk 8]
```diff
@@ -1,71 +1,52 @@
...
-  }
-
-  public joinRoom() {
+  public joinRoom(user: User) {
     return new Promise<{ id: string; board: Board }>((resolve) => {
-      for (let i = 0; i < this.roomsDetails.length; i++) {
-        if (this.roomsDetails[i].users < 3) {
-          this.roomsDetails[i].users++;
+      for (let i = 0; i < this.roomState.length; i++) {
+        if (this.roomState[i].users.length < 3) {
```
Comment (Skrpk) [inline server/game_websocket/room.ts]: `3` is magic number antipattern here. Lest create a variable with name ROOM_MAX_CAPACITY and use it instead.

File: server/game_websocket/room.ts [hunk 9]
```diff
@@ -76,13 +57,24 @@ export class Room {
     });
   }
 
+  public getUsersInRoom(user: User) {
+    // return this.roomsState[user.roomID];
+    const room = this.roomState.filter((room) => {
+      if (room.id === user.roomID) return true;
+    });
+    return room[0].users;
```
Comment (Skrpk) [inline server/game_websocket/room.ts]: I launched app locally, opened few windows with this tab of this app and then tried to reload one of the tabs several tabs. After this the app crashed with this error

```
TypeError: Cannot read properties of undefined (reading 'users')
    at Room.getUsersInRoom (/Users/vs/Documents/projects/germatsch/color-board-game/server/game_websocket/room.ts:65:20)
    at /Users/vs/Documents/projects/germatsch/color-board-game/server/game_websocket/index.ts:53:45
    at Generator.next (<anonymous>)
    at /Users/vs/Documents/projects/germatsch/color-board-game/server/game_websocket/index.ts:8:71
    at new Promise (<anonymous>)
    at __awaiter (/Users/vs/Documents/projects/germatsch/color-board-game/server/game_websocket/index.ts:4:12)
    at Socket.<anonymous> (/Users/vs/Documents/projects/germatsch/color-board-game/server/game_websocket/index.ts:44:40)
    at Socket.emit (node:events:520:28)
    at Socket.emit (node:domain:475:12)
    at Socket.emitReserved (/Users/vs/Documents/projects/germatsch/color-board-game/node_modules/socket.io/dist/typed-events.js:56:22)
[nodemon] app crashed - waiting for file changes before starting...
```

**Comment by zt4ff** [comment]
Hi @Skrpk 

I resolved the changes based on your recommendation and fixed the application with the tests.

You run the test with `yarn run server:test`

Thanks for the awesome feedback

**Comment by Skrpk** [comment]
Hey @zt4ff !
Great job, now it looks much better!
No problems with the code anymore, but now when I launched the app I found this. 
I'm not sure if this is a bug, take a look and check it, please.
https://nimb.ws/odJVl4
I opened one window and started reloading the page and it added more users, but I still have only one.

**Pull request merged**